sw/host/opentitanlib/src/transport/hyperdebug/spi.rs

```diff
@@ -0,0 +1,215 @@
+        })
+    }
+
+    /// Transmit data for a single SPI operation, using one or more USB packets.
+    fn transmit(&self, wbuf: &[u8], rbuf_len: usize) -> Result<()> {
+        let mut buf: [u8; USB_MAX_SIZE] = [0; USB_MAX_SIZE];
+        buf[0..2].clone_from_slice(&USB_SPI_PKT_ID_CMD_TRANSFER_START.to_le_bytes());
+        buf[2..4].clone_from_slice(&(wbuf.len() as u16).to_le_bytes());
+        buf[4..6].clone_from_slice(&(rbuf_len as u16).to_le_bytes());
+        let databytes = std::cmp::min(USB_MAX_SIZE - 6, wbuf.len());
+        buf[6..6 + databytes].clone_from_slice(&wbuf[0..databytes]);
+        self.usb_write_bulk(&buf[0..6 + databytes])?;
+        let mut index = databytes;
+
+        while index < wbuf.len() {
+            buf[0..2].clone_from_slice(&USB_SPI_PKT_ID_CMD_TRANSFER_CONTINUE.to_le_bytes());
+            buf[2..4].clone_from_slice(&(index as u16).to_le_bytes());
+            let databytes = std::cmp::min(USB_MAX_SIZE - 4, wbuf.len() - index);
+            buf[4..4 + databytes].clone_from_slice(&wbuf[index..index+databytes]);
+            self.usb_write_bulk(&buf[0..4 + databytes])?;
+            index += databytes;
+        }
+        Ok(())
+    }
+    
+    /// Receive data for a single SPI operation, using one or more USB packets.
+    fn receive(&self, rbuf: &mut [u8]) -> Result<()> {
+        let mut buf: [u8; USB_MAX_SIZE] = [0; USB_MAX_SIZE];
+        let bytecount = self.usb_read_bulk(&mut buf)?;
+        assert!(buf[0..2] == USB_SPI_PKT_ID_RSP_TRANSFER_START.to_le_bytes());
+        let status_code = u16::from_le_bytes(buf[2..4].try_into()?);
+        assert!(status_code == 0);
+        assert!(bytecount >= 4);
+        let databytes = bytecount - 4;
+        rbuf[0..databytes].clone_from_slice(&buf[4..4+databytes]);
+        let mut index = databytes;
+        while index < rbuf.len() {
+            let bytecount = self.usb_read_bulk(&mut buf)?;
+            assert!(bytecount > 4);
+            assert!(buf[0..2] == USB_SPI_PKT_ID_RSP_TRANSFER_CONTINUE.to_le_bytes());
+            assert!(buf[2..4] == (index as u16).to_le_bytes());
+            let databytes = bytecount - 4;
+            rbuf[index..index+databytes].clone_from_slice(&buf[4..4+databytes]);
+            index += databytes;
+        }
+        Ok(())
+    }
+
+    /// Send one USB packet.
+    fn usb_write_bulk(&self, buf: &[u8]) -> Result<()> {
+        (*self.usb_handle).borrow_mut().write_bulk(self.interface.out_endpoint, buf, TIMEOUT)?;
+        Ok(())
+    }
+
+    /// Receive one USB packet.
+    fn usb_read_bulk(&self, buf: &mut [u8]) -> Result<usize> {
+        Ok((*self.usb_handle).borrow_mut().read_bulk(self.interface.in_endpoint, buf, TIMEOUT)?)
+    }
+}
+
+impl Drop for HyperdebugSpiTarget {
+    fn drop(&mut self) {
+        if let Err(_) = (*self.usb_handle).borrow_mut().release_interface(self.interface.interface) {
+            panic!("Error releasing SPI USB interface");
+        }
+    }
+}
```

Comment:
I don't think you need to perform the release yourself.  It looks like `rusb` does this automatically:

https://docs.rs/rusb/0.9.0/src/rusb/device_handle.rs.html#114

Reply:
I have dropped this section.  Worst case, the OS will release when the process terminates.